src/utility/copy.cpp

```diff
@@ -3,30 +3,134 @@
 extern "C" {
 #include "postgres.h"
 #include "access/table.h"
-#include "catalog/namespace.h"
 #include "commands/copy.h"
-#include "nodes/makefuncs.h"
-#include "parser/parse_coerce.h"
-#include "parser/parse_collate.h"
-#include "parser/parse_expr.h"
-#include "parser/parse_node.h"
+#include "executor/executor.h"
+#include "nodes/parsenodes.h"
 #include "parser/parse_relation.h"
-#include "tcop/tcopprot.h"
-#include "optimizer/optimizer.h"
 #include "utils/rel.h"
 #include "utils/rls.h"
 #include "utils/lsyscache.h"
+#include "parser/scanner.h"
+#include "parser/gram.h"
+#include "tcop/tcopprot.h"
+
+#include "pgduckdb/vendor/pg_ruleutils.h"
+#include "pgduckdb/pgduckdb_ruleutils.h"
 }
 
 #include "pgduckdb/utility/copy.hpp"
-#include "pgduckdb/scan/postgres_scan.hpp"
 #include "pgduckdb/pgduckdb_duckdb.hpp"
-#include "pgduckdb/vendor/pg_list.hpp"
 
 static constexpr char s3_filename_prefix[] = "s3://";
 static constexpr char gcs_filename_prefix[] = "gs://";
 static constexpr char r2_filename_prefix[] = "r2://";
 
+static int
+FindQueryTokenOffset(const char *query, yytokentype token_to_match) {
+	core_yyscan_t yyscanner;
+	core_yy_extra_type yyextra;
+	core_YYSTYPE yylval;
+	YYLTYPE yylloc;
+	int token_offset = -1;
+	int nested_level = 0;
+	int tok;
+
+	yyscanner = scanner_init(query, &yyextra, &ScanKeywords, ScanKeywordTokens);
+
+	for (;;) {
+		tok = core_yylex(&yylval, &yylloc, yyscanner);
+		if (tok == 0) {
+			break;
+		} else if (tok == '(') {
+			nested_level++;
+		} else if (tok == ')') {
+			nested_level--;
+		} else if (tok == token_to_match && nested_level == 0) {
+			token_offset = yylloc;
+			break;
+		}
+	}
+	scanner_finish(yyscanner);
+	return token_offset;
+}
+
+static std::string
+CreateRelationCopyString(ParseState *pstate, CopyStmt *copy_stmt, bool *allowed) {
```

Comment:
A little function comment would be helpful here. Something like this:
```suggestion
/* Returns the relation of the copy_stmt as a fully qualified DuckDB table reference. This is done including the column names if provided in the original copy_stmt, e.g. my_table(column1, column2). This also checks permissions on the table to see if the user is allowed to copy the data from this table */
static std::string
CreateRelationCopyString(ParseState *pstate, CopyStmt *copy_stmt, bool *allowed) {
```